### PR TITLE
Ratelimitpolicy documentation

### PR DESCRIPTION
**Pull request by eguzki**
Added documentation about rate limit policy
* Howto's
* API reference
* User guides
* Implementation designs

Fixes https://github.com/Kuadrant/kuadrant-controller/issues/172

Preview in https://github.com/Kuadrant/kuadrant-controller/tree/ratelimitpolicy-user-guide 

Dev notes: User guides currently not working until https://github.com/Kuadrant/limitador/pull/95 is merged. Meanwhile, Limitador must be restarted to have the new limits reloaded.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -24,281 +33,98 @@ of applications & services when it comes to rate limiting, authentication, autho
...
+### Kuadrant components
 
 | CRD | Description |
 | --- | --- |
-| [RateLimitPolicy](apis/apim/v1alpha1/ratelimitpolicy_types.go) | Enable access control on workloads based on HTTP rate limiting |
-| [AuthPolicy](apis/apim/v1alpha1/authpolicy_types.go) | Enable AuthN and AuthZ based access control on workloads |
+| Control Plane | The control plane takes the customer desired configuration (declaratively as kubernetes custom resources) as input and ensures all components are configured to obey customer's desired behavior.<br> This repository contains the source code of the kuadrant control plane. |
+| [Kuadrant Operator](https://github.com/Kuadrant/kuadrant-operator) | A Kubernetes Operator to manage the lifecycle of the kuadrant deployment |
+| [Authorino](https://github.com/Kuadrant/authorino) | The AuthN/AuthZ enforcer. As the [external istio authorizer](https://istio.io/latest/docs/tasks/security/authorization/authz-custom/) ([envoy external authorization](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/ext_authz_filter) serving [gRPC](https://grpc.io/) service |
```
Comment (guicassolato) [inline README.md:54]:
```suggestion
| [Authorino](https://github.com/Kuadrant/authorino) | The AuthN/AuthZ enforcer. As the [external istio authorizer](https://istio.io/latest/docs/tasks/security/authorization/authz-custom/) ([envoy external authorization](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/ext_authz_filter) serving [gRPC](https://grpc.io/) service) |
```

File: doc/rate-limiting.md [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+* Only supporting HTTPRoute/Gateway references from within the same namespace.
+* `hosts` in rules, `spec.rateLimits[].rules`, do not support wildcard prefixes.
+
+## How: Implementation details
+
+### The WASM Filter
+
+On designing kuadrant rate limiting and considering Istio/Envoy's rate limiting offering,
+we hit two limitations
+([described here](https://docs.google.com/document/d/1ve_8ZBq8TK_wnAZHg69M6-f_q1w-mX4vuP1BC1EuEO8/edit#bookmark=id.5wyq2fj56u94)).
```
Comment (guicassolato) [inline doc/rate-limiting.md:294]: I think this Google Doc is Red Hat internal only.

File: README.md [hunk 3]
```diff
@@ -24,281 +33,98 @@ of applications & services when it comes to rate limiting, authentication, autho
...
+| [Limitador](https://github.com/Kuadrant/limitador) | The external rate limiting service. It exposes a [gRPC](https://grpc.io/) service implementing the [Envoy Rate Limit protocol (v3)](https://www.envoyproxy.io/docs/envoy/latest/api-v3/service/ratelimit/v3/rls.proto). |
```
Comment (guicassolato) [inline README.md:55]: We could avoid the external link to https://grpc.io, IMO.

File: README.md [hunk 4]
```diff
@@ -24,281 +33,98 @@ of applications & services when it comes to rate limiting, authentication, autho
...
+| CRD | Owner | Description |
+| --- | --- | --- |
+| [Kuadrant CRD](https://github.com/Kuadrant/kuadrant-operator/blob/main/api/v1beta1/kuadrant_types.go) | [Kuadrant Operator](https://github.com/Kuadrant/kuadrant-operator) | Represents an instance of kuadrant |
+| [Limitador CRD](doc/ratelimitpolicy-reference.md) | [Limitador Operator](https://github.com/Kuadrant/limitador-operator) | Represents an instance of Limitador |
+| [Authorino CRD](https://github.com/Kuadrant/authorino-operator#the-authorino-custom-resource-definition-crd) | [Authorino Operator](https://github.com/Kuadrant/authorino-operator) | Represents an instance of Authorino |
+| [AuthConfig CRD](https://github.com/Kuadrant/authorino/blob/main/docs/architecture.md#the-authorino-authconfig-custom-resource-definition-crd) | [Authorino](https://github.com/Kuadrant/authorino) | The desired authN and authZ protection for a service |
+
+![architecture](https://i.imgur.com/0vdUZ0l.png)
 
 ## Getting started
```
Comment (guicassolato) [inline README.md:80]: Perhaps a one-liner here saying that the steps are for adding RL-only (no auth)?

File: README.md [hunk 5]
```diff
@@ -24,281 +33,98 @@ of applications & services when it comes to rate limiting, authentication, autho
...
-10. Verify authenticated ratelimit by doing `200 OK` requests 2-3 times.
+* (Optionally) deploy istio ingress gateway using the
+[Gateway](https://gateway-api.sigs.k8s.io/v1alpha2/references/spec/#gateway.networking.k8s.io/v1alpha2.Gateway) resource.
+* Write and apply the Kuadrant's RateLimitPolicy custom resource targeting the Gateway resource.
 
-## Demos
+Your gateway is being rate limited according to the policy you defined.
 
-### [Updating the RateLimitPolicy `targetRef` attribute](/doc/demo-rlp-update-targetref.md)
+## User guide index
```
Comment (guicassolato) [inline README.md:108]:
```suggestion
## User guides
```

File: README.md [hunk 6]
```diff
@@ -24,281 +33,98 @@ of applications & services when it comes to rate limiting, authentication, autho
...
-### [Authenticated rate limiting](/doc/demo-rlp-authenticated.md)
+* [Simple rate limiting for API owners](doc/user-guides/simple-rl-for-api-owners.md)
+* [Authenticated rate limiting for API owners](doc/user-guides/authenticated-rl-for-api-owners.md)
+* [Gateway rate limiting for cluster operators](doc/user-guides/gateway-rl-for-cluster-operators.md)
 
-This demo shows how to configure rate limiting after authentication stage and rate limit configuration
-is per API key basis.
+## [Kuadrant Rate Limiting](/doc/rate-limiting.md)
+
+## FAQ
```
Comment (guicassolato) [inline README.md:118]: Should we omit this section until we have content to add?

File: README.md (hunk 4 above)
Comment (eguzki) [reply README.md:80]: It is supposed to be the "getting started" generic for the kuadrant's protection, not only RL. I will fix that.

Comment (eguzki) [reply README.md:79]: Fixed

File: README.md (hunk 6 above)
Comment (eguzki) [reply README.md]: done

File: README.md (hunk 3 above)
Comment (eguzki) [reply README.md]: link removed

File: README.md [hunk 7]
```diff
@@ -24,281 +32,97 @@ of applications & services when it comes to rate limiting, authentication, autho
...
+[AuthPolicy CRD](apis/apim/v1alpha1/authpolicy_types.go) resources targeting the HTTPRoute resource
+to have your API protected.
 
-## Demos
+### If you are a *Cluster Operator*
 
-### [Updating the RateLimitPolicy `targetRef` attribute](/doc/demo-rlp-update-targetref.md)
+* (Optionally) deploy istio ingress gateway using the
+[Gateway](https://gateway-api.sigs.k8s.io/v1alpha2/references/spec/#gateway.networking.k8s.io/v1alpha2.Gateway) resource.
+* Write and apply the Kuadrant's RateLimitPolicy custom resource targeting the Gateway resource.
```
Comment (guicassolato) [inline README.md:103]:
```suggestion
```

File: README.md [hunk 8]
```diff
@@ -24,281 +32,97 @@ of applications & services when it comes to rate limiting, authentication, autho
...
-curl -v -H 'Host: api.toystore.com' -H 'Authorization: APIKEY BOBKEYFORDEMO' -X POST http://localhost:9080/admin/toy
-```
+### If you are an *API Provider*
 
-10. Verify authenticated ratelimit by doing `200 OK` requests 2-3 times.
+* Deploy the service/API to be protected ("Upstream")
+* Expose the service/API using the kubernetes Gateway API, ie
+[HTTPRoute](https://gateway-api.sigs.k8s.io/v1alpha2/references/spec/#gateway.networking.k8s.io/v1alpha2.HTTPRoute) object.
+* Write and apply the Kuadrant's [RateLimitPolicy](/doc/rate-limiting.md) and/or
+[AuthPolicy CRD](apis/apim/v1alpha1/authpolicy_types.go) resources targeting the HTTPRoute resource
```
Comment (guicassolato) [inline README.md:96]:
```suggestion
[AuthPolicy](apis/apim/v1alpha1/authpolicy_types.go) custom resources targeting the HTTPRoute resource
```

File: README.md [hunk 9]
```diff
@@ -24,281 +32,97 @@ of applications & services when it comes to rate limiting, authentication, autho
...
 
-## Demos
+### If you are a *Cluster Operator*
 
-### [Updating the RateLimitPolicy `targetRef` attribute](/doc/demo-rlp-update-targetref.md)
+* (Optionally) deploy istio ingress gateway using the
+[Gateway](https://gateway-api.sigs.k8s.io/v1alpha2/references/spec/#gateway.networking.k8s.io/v1alpha2.Gateway) resource.
+* Write and apply the Kuadrant's RateLimitPolicy custom resource targeting the Gateway resource.
+* Write and apply the Kuadrant's [RateLimitPolicy](/doc/rate-limiting.md) and/or
+[AuthPolicy CRD](apis/apim/v1alpha1/authpolicy_types.go) resources targeting the Gateway resource
```
Comment (guicassolato) [inline README.md:105]:
```suggestion
[AuthPolicy](apis/apim/v1alpha1/authpolicy_types.go) custom resources targeting the Gateway resource
```

File: doc/rate-limiting.md (hunk 2 above)
Comment (eguzki) [reply doc/rate-limiting.md]: fixed

File: README.md (hunk 7 above)
Comment (eguzki) [reply README.md:103]: I have updated that. This suggested change does not apply.

File: README.md (hunk 3 above)
Comment (guicassolato) [reply README.md]: The other one in the Authorino description intentionally kept?

File: README.md (hunk 7 above)
Comment (guicassolato) [reply README.md:103]: The way I read it, it looked like the line was repeated. I guess what you want to capture here then is the difference between infrastructure RL and business level RL/auth rules maybe, both applied at the level of the Gateway?

File: README.md (hunk 3 above)
Comment (eguzki) [reply README.md]: fixed

File: README.md (hunk 7 above)
Comment (eguzki) [reply README.md]: much simpler than that. The line was leaked. I have removed it

**Review by guicassolato — APPROVED** [review]

**Pull request merged**